activation/atxdb.go

```diff
@@ -87,13 +89,66 @@ func (m *ActivationDb) CalcActiveSetFromView(a *types.ActivationTx) (uint32, err
 		mp[blk] = struct{}{}
 	}
 
-	m.meshDb.ForBlockInView(mp, firstLayerOfLastEpoch, traversalFunc, errHandler)
+	db.meshDb.ForBlockInView(mp, firstLayerOfLastEpoch, traversalFunc, errHandler)
 	activesetCache.Add(common.BytesToHash(bytes), counter)
 
 	return counter, nil
 
 }
 
+//todo: move to config
+const NIPST_LAYERTIME = 1000
+
+func (db *ActivationDb) ValidateAtx(atx *types.ActivationTx) error {
+	// validation rules: no other atx with same id and sequence number
+	// if s != 0 the prevAtx is valid and it's seq num is s -1
+	// positioning atx is valid
+	// validate nipst duration?
+	// fields 1-7 of the atx are the challenge of the poet
+	// layer index i^ satisfies i -i^ < (layers_passed during nipst creation) ANTON: maybe should be ==?
+	// the atx view contains d active Ids
+
+	eatx, _ := db.GetAtx(atx.Id())
+	if eatx != nil {
+		return fmt.Errorf("found atx with same id")
+	}
+	prevAtxIds, err := db.GetNodeAtxIds(atx.NodeId)
+	if err == nil && len(prevAtxIds) > 0 {
+		prevAtx, err := db.GetAtx(prevAtxIds[len(prevAtxIds)-1])
+		if err == nil {
+			if prevAtx.Sequence >= atx.Sequence {
+				return fmt.Errorf("found atx with same seq or greater")
+			}
+		}
+	} else {
+		if prevAtxIds == nil && atx.PrevATXId != types.EmptyAtx {
```

Comment:
Replace `prevAtxIds == nil` with `len(prevAtxIds) == 0`?

This would better convey the intention and also be safer (what If someone makes `db.GetNodeAtxIds` return an empty slice instead of `nil` when nothing is found?).

Reply:
More importantly - you're not checking for the opposite case, where someone **doesn't** link to a previous ATX, but one exists.

Reply:
BTW, this is the biggest thing we need to defend against, since it allows working on multiple ATXs concurrently.